feat: robots directive, full robots tag support (#12998)
shopware/platform: changelog/_unreleased/2025-10-14-support-all-robots-txt-directives.md

author_github: BrocksiNet
---
# Core
* Added `Shopware\Storefront\Page\Robots\Parser\RobotsDirectiveParser` to parse robots.txt configurations
* Added `Shopware\Storefront\Page\Robots\Parser\ParsedRobots` value object for parsed robots.txt data
* Added `Shopware\Storefront\Page\Robots\Struct\RobotsDirective` value object representing a single directive
* Added `Shopware\Storefront\Page\Robots\Struct\RobotsUserAgentBlock` value object representing a user-agent block
* Added `globalUserAgentBlocks` property to `Shopware\Storefront\Page\Robots\RobotsPage` to store global user-agent blocks
* Changed `Shopware\Storefront\Page\Robots\RobotsPageLoader` to parse and separate global user-agent blocks from domain-specific path rules
* Changed `Shopware\Storefront\Page\Robots\Struct\DomainRuleStruct` to support new parser while maintaining backward compatibility
* Added new system configuration option `core.basicInformation.robotsDisableDefaults` to disable Shopware's default robots.txt rules

___
# Storefront
* Changed rendering of robots.txt to support custom User-agent blocks
* Changed robots.txt template to respect new `robotsDisableDefaults` configuration

___
# Upgrade Information

> ## Technical Changes (Backward Compatible)

Reviewer: remember that this will be the headline in the upgrade file. 

`Technical Changes (Backward Compatible)` will be quite lost without its context. 

so probably, you should add a title for the whole topic and all `##` headlines should become `###` headlines

The constructor signature of `Shopware\Storefront\Page\Robots\Struct\DomainRuleStruct` has been updated to accept both `string` and `ParsedRobots` objects:

```php
// Old signature (still works)
new DomainRuleStruct('Disallow: /admin/', '/en');

// New signature (also works)

> new DomainRuleStruct($parsedRobots, '/en');

Reviewer: maybe add it like this and some example parameters?
```suggestion
$parsedRobots = new ParsedRobots();
new DomainRuleStruct($parsedRobots, '/en');
```

```

> **This change is backward compatible** - all existing code passing a string continues to work without modification. The union type `ParsedRobots|string` accepts everything the old `string` type accepted, plus the new parsed object type for internal use.

Reviewer: I don't think we need to explain that again. you already state above, that you accept both types

---
title: Support all robots.txt directives including User-agent blocks
issue: NEXT-12787
author: Björn Meyer
author_email: b.meyer@shopware.com
author_github: BrocksiNet
---
# Core

> * Added `Shopware\Storefront\Page\Robots\Parser\RobotsDirectiveParser` to parse robots.txt with error tracking and extensibility events

Reviewer: ```suggestion
* Added `Shopware\Storefront\Page\Robots\Parser\RobotsDirectiveParser` to parse `robots.txt` file with error tracking and extensibility events
```

* Added `Shopware\Storefront\Page\Robots\Event\RobotsDirectiveParsingEvent` to allow modification of parsed results
* Added `Shopware\Storefront\Page\Robots\Event\RobotsUnknownDirectiveEvent` to handle custom directives

> * Added robots.txt parsing value objects: `ParsedRobots`, `ParseIssue`, `ParseIssueSeverity`, `RobotsDirective`, `RobotsDirectiveType`, `RobotsUserAgentBlock`

Reviewer: probably not relevant to mention. those are internals which are probably not used directly by externals

> * Added `Shopware\Storefront\Page\Robots\RobotsConfigChangeSubscriber` to log parsing issues when robots.txt is saved

Reviewer: probably also not really relevant for externals

> * Added system configuration option `core.basicInformation.robotsDisableDefaults` to disable default robots.txt rules

Reviewer: instead of mentioning a new config with the technical name I would add something to the upgrade information, mentioning a new option in the admin

Reviewer: I just saw, that this is already mentioned, so this can be removed.

* Deprecated passing a string to `Shopware\Storefront\Page\Robots\Struct\DomainRuleStruct` constructor - use `ParsedRobots` object instead

___
# Storefront

> * Changed rendering of robots.txt to support custom User-agent blocks

Reviewer: ```suggestion
* Changed rendering of `robots.txt` file to support custom User-agent blocks
```

> * Changed robots.txt template to respect new `robotsDisableDefaults` configuration

Reviewer: imo, also not really relevant

___
# Upgrade Information

> ## robots.txt Configuration Enhancement

Reviewer: the documentation of the new feature is very extensive. which is good, don't get me wrong. 
but maybe it would be better to prepare a documentation article first and just link it here in the changelog. wdyt?

Reviewer 2: Documentation should be done. But in terms of the new changelog process this should also go into the release notes, right?

The robots.txt system has been enhanced to support the full robots.txt standard including User-agent blocks and all common directives.

### Technical Changes

The constructor of `Shopware\Storefront\Page\Robots\Struct\DomainRuleStruct` now accepts `ParsedRobots` objects for proper parsing with validation:

```php
// ⚠️ DEPRECATED (will be removed in v6.8.0): String format
new DomainRuleStruct('Disallow: /admin/', '/en');

// ✅ NEW: Use ParsedRobots for proper parsing with validation and events
$parser = new RobotsDirectiveParser($this->eventDispatcher);
$parsedRobots = $parser->parse("
    User-agent: Googlebot
    Crawl-delay: 10
    Disallow: /admin/
", $context);
new DomainRuleStruct($parsedRobots, '/en');
```

> **Migration required**: The string format is deprecated and will be removed in v6.8.0. Use `RobotsDirectiveParser::parse()` instead to get validation, error logging, and event support.

Reviewer: ```suggestion
Passing a string to the `DomainRuleStruct` is deprecated and will be removed with the next major version.
Use `RobotsDirectiveParser::parse()` instead to get validation, error logging, and event support.
```

### Full robots.txt directive support

The robots.txt configuration now supports the full robots.txt standard including:

#### User-agent blocks
You can now define custom User-agent blocks with specific directives:

```
User-agent: Googlebot
Crawl-delay: 10
Disallow: /admin/

User-agent: Bingbot
Disallow: /secret/
```

#### Global vs Domain-specific rules

- **User-agent blocks** are global: They are collected from all sales channels, deduplicated, and rendered once
- **Path directives** (Allow/Disallow) within user-agent blocks are domain-specific: They get the domain's base path applied automatically

#### Example: Multi-language shop

If you have a shop with English and German sales channels at `/en` and `/de`:

**English sales channel config:**
```
User-agent: Googlebot
Crawl-delay: 10
Disallow: /account/
```

**German sales channel config:**
```
User-agent: Googlebot
Crawl-delay: 10
Disallow: /konto/
```

**Resulting robots.txt:**
```
User-agent: *
Allow: /
Disallow: /*?

> Allow: /*theme/

Reviewer: where does this media stuff come from? seems unrelated and is a bit confusing

#### Supported directives

The following robots.txt directives are now supported:
- `User-agent`: Define which crawler the rules apply to
- `Allow`: Allow crawling of specific paths
- `Disallow`: Disallow crawling of specific paths
- `Crawl-delay`: Request crawl rate limit
- `Sitemap`: Specify sitemap URL
- `Request-rate`: Alternative rate limiting
- `Visit-time`: Specify crawl time windows
- `Host`: Specify preferred domain

#### Disable default rules

You can now disable Shopware's default robots.txt rules via the new configuration option "Disable default robots.txt rules" in Settings > Basic information. This gives you full control over the robots.txt content.

#### Parsing error detection and logging

When you save robots.txt configuration in the admin, Shopware now validates the syntax and logs any issues:

**Error types:**
- **Errors**: Critical issues like malformed lines (missing colon separator)
- **Warnings**: Non-critical issues like unknown directives or directives outside user-agent blocks

**Logging behavior:**
- Issues are logged only when configuration is saved via admin (not on every page load)
- Each issue includes line number, line content, and error reason

> - Errors are logged with `error` level, warnings with `warning` level

Reviewer: ![](https://i.kym-cdn.com/entries/icons/original/000/000/015/oreally.jpg)

**Example**: Supporting Yandex's `Clean-param` directive:
```php
#[AsEventListener(event: RobotsUnknownDirectiveEvent::class)]
class YandexDirectiveSubscriber
{
    public function onUnknownDirective(RobotsUnknownDirectiveEvent $event): void
    {
        if ($event->getDirectiveType() === 'Clean-param') {
            // Mark as handled so it doesn't generate a warning
            $event->setHandled(true);
        }
    }
}
```

**Example**: Adding custom validation to enforce security policies:
```php
#[AsEventListener(event: RobotsDirectiveParsingEvent::class)]
class RobotsSecurityPolicySubscriber
{
    public function onParsingComplete(RobotsDirectiveParsingEvent $event): void
    {
        $parsed = $event->getParsedResult();

        // Ensure /admin/ is always disallowed for all bots
        $hasAdminDisallow = false;
        foreach ($parsed->orphanedPathDirectives as $directive) {
            if ($directive->type === RobotsDirectiveType::DISALLOW && $directive->value === '/admin/') {
                $hasAdminDisallow = true;
                break;
            }
        }

        if (!$hasAdminDisallow) {
            // Add a warning about missing security directive

>             $newIssues = [

Reviewer: that is not really good dev ex :thinking: 

should be more something like this: 
```
$event->parsedResult->issues->add(new ParseIssue(0, '', 'Security: /admin/ should be disallowed', ParseIssueSeverity::WARNING))
```
